src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/ViewDataDictionary.cs

```diff
@@ -395,37 +455,54 @@ public ViewDataInfo GetViewDataInfo(string expression)
             return ViewDataEvaluator.Eval(this, expression);
         }
 
-        // This method will execute before the derived type's instance constructor executes. Derived types must
-        // be aware of this and should plan accordingly. For example, the logic in SetModel() should be simple
-        // enough so as not to depend on the "this" pointer referencing a fully constructed object.
+        /// <summary>
+        /// Set <see cref="ModelExplorer"/> to ensure <see cref="Model"/> and <see cref="ModelExplorer.Model"/>
+        /// reflect the new <paramref name="value"/>.
+        /// </summary>
+        /// <param name="value">New <see cref="Model"/> value.</param>
         protected virtual void SetModel(object value)
         {
-            EnsureCompatible(value);
-
-            // Reset or override ModelMetadata based on runtime value type. Fall back to declared type if value is
-            // null. When called from a constructor, current ModelExplorer may already be set to preserve
-            // customizations made in parent scope. But ModelExplorer is never null after instance is initialized.
-            var modelType = GetModelType(value);
-            Type metadataModelType = null;
-            if (ModelExplorer != null)
+            // Update ModelExplorer to reflect the new value. When possible, preserve ModelMetadata to avoid losing
+            // property information.
+            var modelType = value?.GetType();
+            if (ModelMetadata.MetadataKind == ModelMetadataKind.Type &&
```

Comment:
Thought about a `SetModelCore()` method but `SetModel()` and the main copy constructor each have unique branches in the logic. [One block](https://github.com/aspnet/Mvc/pull/4403/files/5fee305a5ddf8a0206e1945130290af389b4044f#diff-dd434cfe5fb0b1b08c4bdac8467917beR236) in that constructor could be done instead of calling `SetModelCore()` -- that's basically another branch in a large '||' condition. And could pass a `bool` into `SetModelCore()` for the [block below](https://github.com/aspnet/Mvc/pull/4403/files/5fee305a5ddf8a0206e1945130290af389b4044f#diff-dd434cfe5fb0b1b08c4bdac8467917beR497) if you feel strongly about it. LMK.